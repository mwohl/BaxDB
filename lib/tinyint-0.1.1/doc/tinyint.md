tinyint 0.1.1
=============

Synopsis
--------

    % CREATE EXTENSION tinyint;
    CREATE EXTENSION

    % SELECT 1::tinyint;


Description
-----------

This module provides 1 byte signed integer type with operators,
index access methods and support functions. You can use it as
like the corresponding data type in other RDBMS as well as
smallint, integer or bigint in PostgreSQL. Although the byte
used for 1 tinyint is exactly 1 byte, you need to take account
into its position in the row if you really care the table size
since PostgreSQL usually pads extra bytes for the alignment of 4.

tinyint holds -127 to 127. +, -, *, %. min(), max(), sum(), avg(),
btree index and gin index are supported as other integer types.

Support
-------

This library is stored in an open [GitHub
repository](http://github.com/umitanuki/tinyint-postgresql).
Feel free to fork and contribute! Please file bug reports
via [GitHub Issues](http://github.com/umitanuki/tinyint-postgresql/issues/).

Author
------

[Hitoshi Harada](mailto:umi.tanuki@gmail.com)

Copyright and License
---------------------

Copyright (c) Hitoshi Harada

This module is free software; you can redistribute it and/or modify it under
the [PostgreSQL License](http://www.opensource.org/licenses/postgresql).

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose, without fee, and without a written agreement is
hereby granted, provided that the above copyright notice and this paragraph
and the following two paragraphs appear in all copies.

In no event shall Hitoshi Harada be liable to any party for direct,
indirect, special, incidental, or consequential damages, including
lost profits, arising out of the use of this software and its documentation,
even if Hitoshi Harada has been advised of the possibility of such damage.

Hitoshi Harada specifically disclaim any warranties,
including, but not limited to, the implied warranties of merchantability and 
fitness for a particular purpose. The software provided hereunder is on an "as 
is" basis, and Hitoshi Harada has no obligations to provide maintenance,
support, updates, enhancements, or modifications.

