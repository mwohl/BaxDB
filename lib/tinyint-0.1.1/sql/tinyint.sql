SET client_min_messages = warning;
\set ECHO none
\i tinyint.sql
\set ECHO all
RESET client_min_messages;

SELECT 0::tinyint;
SELECT 1::integer::tinyint::smallint::tinyint::bigint::tinyint;

SELECT 1::tinyint < 2::tinyint;
SELECT -1::tinyint > +10::tinyint;
SELECT 1::smallint > 0::tinyint;
SELECT 1::integer > 0::tinyint;
SELECT 1::bigint > 0::tinyint;
SELECT 1::tinyint < 0::smallint;
SELECT 1::tinyint < 0::integer;
SELECT 1::tinyint < 0::bigint;

SELECT 1::tinyint + 1::tinyint;
SELECT 1::tinyint + 1::smallint;
SELECT 1::tinyint + 1::integer;
SELECT 1::tinyint + 1::bigint;
SELECT 1::smallint + 1::tinyint;
SELECT 1::integer + 1::tinyint;
SELECT 1::bigint + 1::tinyint;
SELECT 2::tinyint * 2::tinyint;
SELECT 4::tinyint / -2::tinyint;
SELECT 2::smallint * 2::tinyint;
SELECT 2::integer * 2::tinyint;
SELECT 2::bigint * 2::tinyint;
SELECT 2::smallint / 2::tinyint;
SELECT 2::integer / 2::tinyint;
SELECT 2::bigint / 2::tinyint;
SELECT 2::tinyint * 2::smallint;
SELECT 2::tinyint * 2::integer;
SELECT 2::tinyint * 2::bigint;
SELECT 2::tinyint / 2::smallint;
SELECT 2::tinyint / 2::integer;
SELECT 2::tinyint / 2::bigint;

SELECT MAX(val), MIN(val), AVG(val), SUM(val) FROM (VALUES('1'::tinyint),('2'),('5'),('10')) AS s(val);

CREATE TABLE t_tinyint (vals tinyint[]);
INSERT INTO t_tinyint VALUES('{0,0,0,1,2,3,0}'),('{1,2,100,127,-126}'::tinyint[]);
CREATE INDEX t_tinyint_gin ON t_tinyint USING gin(vals);

SELECT 128::tinyint;
SELECT 16::tinyint * 16::tinyint;
SELECT 127::tinyint * 32767::smallint;
SELECT 16::tinyint / 0;
SELECT (1::bigint << 63) * 2::tinyint;
