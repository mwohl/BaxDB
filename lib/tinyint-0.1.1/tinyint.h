
#ifndef TINYINT_H
#define TINYINT_H

#include <ctype.h>

#include "fmgr.h"


typedef char tinyint;

#define DatumGetTinyInt(X) ((tinyint) GET_1_BYTE(X))
#define TinyIntGetDatum(X) ((Datum) SET_1_BYTE(X))

#define PG_GETARG_TINYINT(n)	DatumGetTinyInt(PG_GETARG_DATUM(n))
#define PG_RETURN_TINYINT(x)	return TinyIntGetDatum(x)

extern Datum tinyint_in(PG_FUNCTION_ARGS);
extern Datum tinyint_out(PG_FUNCTION_ARGS);
extern Datum tinyint_recv(PG_FUNCTION_ARGS);
extern Datum tinyint_send(PG_FUNCTION_ARGS);

extern Datum tinyint_to_i2(PG_FUNCTION_ARGS);
extern Datum i2_to_tinyint(PG_FUNCTION_ARGS);
extern Datum tinyint_to_i4(PG_FUNCTION_ARGS);
extern Datum i4_to_tinyint(PG_FUNCTION_ARGS);
extern Datum tinyint_to_i8(PG_FUNCTION_ARGS);
extern Datum i8_to_tinyint(PG_FUNCTION_ARGS);
extern Datum tinyint_to_bool(PG_FUNCTION_ARGS);
extern Datum bool_to_tinyint(PG_FUNCTION_ARGS);

extern Datum tinyint_eq(PG_FUNCTION_ARGS);
extern Datum tinyint_ne(PG_FUNCTION_ARGS);
extern Datum tinyint_lt(PG_FUNCTION_ARGS);
extern Datum tinyint_le(PG_FUNCTION_ARGS);
extern Datum tinyint_gt(PG_FUNCTION_ARGS);
extern Datum tinyint_ge(PG_FUNCTION_ARGS);

extern Datum tinyint_i2_eq(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_ne(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_lt(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_le(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_gt(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_ge(PG_FUNCTION_ARGS);

extern Datum tinyint_i4_eq(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_ne(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_lt(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_le(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_gt(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_ge(PG_FUNCTION_ARGS);

extern Datum tinyint_i8_eq(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_ne(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_lt(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_le(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_gt(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_ge(PG_FUNCTION_ARGS);

extern Datum i2_tinyint_eq(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_ne(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_lt(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_le(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_gt(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_ge(PG_FUNCTION_ARGS);

extern Datum i4_tinyint_eq(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_ne(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_lt(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_le(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_gt(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_ge(PG_FUNCTION_ARGS);

extern Datum i8_tinyint_eq(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_ne(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_lt(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_le(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_gt(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_ge(PG_FUNCTION_ARGS);

extern Datum tinyint_um(PG_FUNCTION_ARGS);
extern Datum tinyint_up(PG_FUNCTION_ARGS);
extern Datum tinyint_pl(PG_FUNCTION_ARGS);
extern Datum tinyint_mi(PG_FUNCTION_ARGS);
extern Datum tinyint_mul(PG_FUNCTION_ARGS);
extern Datum tinyint_div(PG_FUNCTION_ARGS);
extern Datum tinyint_inc(PG_FUNCTION_ARGS);

extern Datum tinyint_i2_pl(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_mi(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_mul(PG_FUNCTION_ARGS);
extern Datum tinyint_i2_div(PG_FUNCTION_ARGS);

extern Datum tinyint_i4_pl(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_mi(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_mul(PG_FUNCTION_ARGS);
extern Datum tinyint_i4_div(PG_FUNCTION_ARGS);

extern Datum tinyint_i8_pl(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_mi(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_mul(PG_FUNCTION_ARGS);
extern Datum tinyint_i8_div(PG_FUNCTION_ARGS);

extern Datum i2_tinyint_pl(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_mi(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_mul(PG_FUNCTION_ARGS);
extern Datum i2_tinyint_div(PG_FUNCTION_ARGS);

extern Datum i4_tinyint_pl(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_mi(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_mul(PG_FUNCTION_ARGS);
extern Datum i4_tinyint_div(PG_FUNCTION_ARGS);

extern Datum i8_tinyint_pl(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_mi(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_mul(PG_FUNCTION_ARGS);
extern Datum i8_tinyint_div(PG_FUNCTION_ARGS);

extern Datum tinyint_mod(PG_FUNCTION_ARGS);
extern Datum tinyint_abs(PG_FUNCTION_ARGS);
extern Datum tinyint_larger(PG_FUNCTION_ARGS);
extern Datum tinyint_smaller(PG_FUNCTION_ARGS);
extern Datum tinyint_and(PG_FUNCTION_ARGS);
extern Datum tinyint_or(PG_FUNCTION_ARGS);
extern Datum tinyint_xor(PG_FUNCTION_ARGS);
extern Datum tinyint_shl(PG_FUNCTION_ARGS);
extern Datum tinyint_shr(PG_FUNCTION_ARGS);
extern Datum tinyint_not(PG_FUNCTION_ARGS);

extern Datum bt_tinyint_cmp(PG_FUNCTION_ARGS);
extern Datum bt_tinyint_i2_cmp(PG_FUNCTION_ARGS);
extern Datum bt_tinyint_i4_cmp(PG_FUNCTION_ARGS);
extern Datum bt_tinyint_i8_cmp(PG_FUNCTION_ARGS);
extern Datum bt_i2_tinyint_cmp(PG_FUNCTION_ARGS);
extern Datum bt_i4_tinyint_cmp(PG_FUNCTION_ARGS);
extern Datum bt_i8_tinyint_cmp(PG_FUNCTION_ARGS);

extern Datum hash_tinyint(PG_FUNCTION_ARGS);

extern Datum tinyint_sum(PG_FUNCTION_ARGS);
extern Datum tinyint_avg_accum(PG_FUNCTION_ARGS);

#endif   /* TINYINT_H */
