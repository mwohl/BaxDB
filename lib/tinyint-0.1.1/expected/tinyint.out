SET client_min_messages = warning;
\set ECHO none
RESET client_min_messages;
SELECT 0::tinyint;
 tinyint 
---------
 0
(1 row)

SELECT 1::integer::tinyint::smallint::tinyint::bigint::tinyint;
 tinyint 
---------
 1
(1 row)

SELECT 1::tinyint < 2::tinyint;
 ?column? 
----------
 t
(1 row)

SELECT -1::tinyint > +10::tinyint;
 ?column? 
----------
 f
(1 row)

SELECT 1::smallint > 0::tinyint;
 ?column? 
----------
 t
(1 row)

SELECT 1::integer > 0::tinyint;
 ?column? 
----------
 t
(1 row)

SELECT 1::bigint > 0::tinyint;
 ?column? 
----------
 t
(1 row)

SELECT 1::tinyint < 0::smallint;
 ?column? 
----------
 f
(1 row)

SELECT 1::tinyint < 0::integer;
 ?column? 
----------
 f
(1 row)

SELECT 1::tinyint < 0::bigint;
 ?column? 
----------
 f
(1 row)

SELECT 1::tinyint + 1::tinyint;
 ?column? 
----------
 2
(1 row)

SELECT 1::tinyint + 1::smallint;
 ?column? 
----------
        2
(1 row)

SELECT 1::tinyint + 1::integer;
 ?column? 
----------
        2
(1 row)

SELECT 1::tinyint + 1::bigint;
 ?column? 
----------
        2
(1 row)

SELECT 1::smallint + 1::tinyint;
 ?column? 
----------
        2
(1 row)

SELECT 1::integer + 1::tinyint;
 ?column? 
----------
        2
(1 row)

SELECT 1::bigint + 1::tinyint;
 ?column? 
----------
        2
(1 row)

SELECT 2::tinyint * 2::tinyint;
 ?column? 
----------
 4
(1 row)

SELECT 4::tinyint / -2::tinyint;
 ?column? 
----------
 -2
(1 row)

SELECT 2::smallint * 2::tinyint;
 ?column? 
----------
        4
(1 row)

SELECT 2::integer * 2::tinyint;
 ?column? 
----------
        4
(1 row)

SELECT 2::bigint * 2::tinyint;
 ?column? 
----------
        4
(1 row)

SELECT 2::smallint / 2::tinyint;
 ?column? 
----------
        1
(1 row)

SELECT 2::integer / 2::tinyint;
 ?column? 
----------
        1
(1 row)

SELECT 2::bigint / 2::tinyint;
 ?column? 
----------
        1
(1 row)

SELECT 2::tinyint * 2::smallint;
 ?column? 
----------
        4
(1 row)

SELECT 2::tinyint * 2::integer;
 ?column? 
----------
        4
(1 row)

SELECT 2::tinyint * 2::bigint;
 ?column? 
----------
        4
(1 row)

SELECT 2::tinyint / 2::smallint;
 ?column? 
----------
        1
(1 row)

SELECT 2::tinyint / 2::integer;
 ?column? 
----------
        1
(1 row)

SELECT 2::tinyint / 2::bigint;
 ?column? 
----------
        1
(1 row)

SELECT MAX(val), MIN(val), AVG(val), SUM(val) FROM (VALUES('1'::tinyint),('2'),('5'),('10')) AS s(val);
 max | min |        avg         | sum 
-----+-----+--------------------+-----
 10  | 1   | 4.5000000000000000 |  18
(1 row)

CREATE TABLE t_tinyint (vals tinyint[]);
INSERT INTO t_tinyint VALUES('{0,0,0,1,2,3,0}'),('{1,2,100,127,-126}'::tinyint[]);
CREATE INDEX t_tinyint_gin ON t_tinyint USING gin(vals);
SELECT 128::tinyint;
ERROR:  tinyint out of range
SELECT 16::tinyint * 16::tinyint;
ERROR:  tinyint out of range
SELECT 127::tinyint * 32767::smallint;
ERROR:  smallint out of range
SELECT 16::tinyint / 0;
ERROR:  division by zero
SELECT (1::bigint << 63) * 2::tinyint;
ERROR:  bigint out of range
